firedrake/slate/slac/tsfc_driver.py

```diff
@@ -47,10 +60,12 @@ def compile_terminal_form(tensor, prefix=None, tsfc_parameters=None):
     cxt_kernels = []
     for orig_it_type, integrals in transformed_integrals.items():
         subkernel_prefix = prefix + "%s_to_" % orig_it_type
-        kernels = tsfc_compile(Form(integrals),
+        form = Form(integrals)
+        kernels = tsfc_compile(form,
                                subkernel_prefix,
                                parameters=tsfc_parameters)
         cxt_k = ContextKernel(tensor=tensor,
```

Comment:
So I think (discussed) that what you want is a bucket which contains tsfc kernels that correspond to the different integral types.  So why not make that?

Then the coefficient_mapping you need can be created by composing the coefficient mapping from the tensor's coefficients to the form that provides each of these integral type's coefficients with the tsfc kernel's coefficient map.  No need to split the form (which is good, because you're no longer repeating work).